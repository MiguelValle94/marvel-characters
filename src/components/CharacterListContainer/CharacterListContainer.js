import React, { useEffect } from 'react';
import { connect } from 'react-redux';

import CharacterList from './CharacterList';
import FiltersContainer from './FiltersContainer';
import CharactersPagination from './CharactersPagination';
import { fetchCharacters } from '../../state/actions/characters.actions';
import { selectPage } from '../../state/actions/pagination.actions';

function CharacterListContainer({
  characters,
  isLoading,
  dispatch,
  filters,
  pagination
}) {
  const { orderBy, nameStartsWith } = filters;
  const { page, pages, limit } = pagination;

  useEffect(() => {
    dispatch(fetchCharacters(orderBy, nameStartsWith, page, limit));
  }, [dispatch, orderBy, nameStartsWith, page, limit]);

  function handleSelectPage(page) {
    dispatch(selectPage(page));
  }

  return (
    <div className='container'>
      <FiltersContainer />
      <CharactersPagination
        page={page}
        pages={pages}
        onSelectPage={handleSelectPage}
      />
      {isLoading ? (
        <p>Loading...</p>
      ) : (
        <CharacterList characters={characters} />
      )}
      <CharactersPagination
        page={page}
        pages={pages}
        onSelectPage={handleSelectPage}
      />
    </div>
  );
}

const mapStateToProps = (state) => ({
  characters: state.characters.characters,
  isLoading: state.characters.isLoading,
  filters: state.filters,
  pagination: state.pagination
});

export default connect(mapStateToProps)(CharacterListContainer);
