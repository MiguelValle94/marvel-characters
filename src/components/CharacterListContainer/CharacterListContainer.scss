@import '../../styles/variables.scss';

.characters-list {
  &__total {
    align-items: center;
    display: flex;
    padding: var(--space);
  }

  &__row {
    display: flex;
    justify-content: flex-end;

    &--space-between {
      justify-content: space-between;
    }

    @media only screen and (max-width: 767px) {
      flex-direction: column-reverse;
    }
  }
}
