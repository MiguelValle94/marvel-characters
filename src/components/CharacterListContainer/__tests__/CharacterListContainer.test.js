import React from 'react';
import { mount } from 'enzyme';

import CharacterListContainer from '../CharacterListContainer';
import { mockStore } from '../../../state/__mocks__/store';

jest.mock('../CharacterList', () => () => <></>);
jest.mock('../FiltersContainer', () => () => <></>);
jest.mock('../CharactersPagination', () => () => <></>);

const setUpMount = (initialState) => {
  const store = mockStore(initialState);
  const component = mount(<CharacterListContainer store={store} />);
  return component;
};

describe('CharacterListContainer', () => {
  let component;

  it('should display loading message while loading', () => {
    component = setUpMount({
      characters: { isLoading: true },
      filters: { orderBy: 'name' },
      pagination: { page: 1, pages: 2 }
    });
    expect(component).toMatchSnapshot();
  });

  it('should display characters list', () => {
    component = setUpMount({
      characters: { characters: [{}, {}] },
      filters: { orderBy: 'name' },
      pagination: { page: 1, pages: 2 }
    });
    expect(component).toMatchSnapshot();
  });
});
