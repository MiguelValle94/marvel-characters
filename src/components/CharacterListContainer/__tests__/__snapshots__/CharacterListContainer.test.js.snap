// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterListContainer should display characters list 1`] = `
<Connect(CharacterListContainer)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CharacterListContainer
    characters={
      Array [
        Object {},
        Object {},
      ]
    }
    dispatch={[Function]}
    filters={
      Object {
        "orderBy": "name",
      }
    }
    pagination={
      Object {
        "page": 1,
        "pages": 2,
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <div
      className="container"
    >
      <Component />
      <Component
        onSelectPage={[Function]}
        page={1}
        pages={2}
      />
      <Component
        characters={
          Array [
            Object {},
            Object {},
          ]
        }
      />
      <Component
        onSelectPage={[Function]}
        page={1}
        pages={2}
      />
    </div>
  </CharacterListContainer>
</Connect(CharacterListContainer)>
`;

exports[`CharacterListContainer should display loading message while loading 1`] = `
<Connect(CharacterListContainer)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CharacterListContainer
    dispatch={[Function]}
    filters={
      Object {
        "orderBy": "name",
      }
    }
    isLoading={true}
    pagination={
      Object {
        "page": 1,
        "pages": 2,
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <div
      className="container"
    >
      <Component />
      <Component
        onSelectPage={[Function]}
        page={1}
        pages={2}
      />
      <p>
        Loading...
      </p>
      <Component
        onSelectPage={[Function]}
        page={1}
        pages={2}
      />
    </div>
  </CharacterListContainer>
</Connect(CharacterListContainer)>
`;
